exercises/joseph_torsney/ex01_basics/src/main.cpp

```diff
@@ -0,0 +1,129 @@
+#include <iostream> 
+#include <fstream>
+#include <vector>
+#include <map>
+#include <algorithm>
+#include <string>
+#include <sstream>
+
+/** Checks whether a character is punctuation
+ * 
+ * @param c 
+ * @return true if c is punctuation
+ * @return false otherwise
+ */
+bool myIsPunct(char c) {
+  return(
+    c == '.' ||
+    c == ',' ||
+    c == '?' ||
+    c == '\'' ||
+    c == '\"' ||
+    c == '!' ||
+    c == ')' ||
+    c == '('
+  );
+}
+
+/** Transforms a string to lower case.
+ * 
+ * @param str the string to transform
+ * @return std::string the string in lower case
+ */
+std::string toLowerStr(std::string str) {
+  std::transform(str.begin(), str.end(), str.begin(), 
+    [](unsigned char c){ return std::tolower(c); }
+  );
+
+  return str;
+}
+
+/** Removes punctuation from a string
+ * 
+ * @param str the string to process
+ * @return std::string with no punctuation
+ */
+std::string removePunct(std::string str)
```

Comment:
Please can you pass the parameter as const ref ie const std::string&

Reply:
How do I do that here, since I'm calling .erase on str?

Reply:
I think what I meant to say was that it would be more efficient to change the function signature to:

void removePunct(std::string& str)

Same adjustment needed to my other comment on toLowerStr.

If you do these two changes you won't be able to nest the calls to these 2 functions any more where you manipulate wordWithoutHyphens so you gain a line of code. But you won't have any string copies happening which means the code will be faster so I think that's worth it.

At this point you could do away with the two functions entirely and just put the contents of removePunct\toLowerStr in wordUsageFromFile. However the std::transform and std::erase syntax is quite dense so having the functions makes it easy to read.